File: c/medium/leetcode_5.c
//
// Created by ywh on 2020/7/22.
//
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#define max(a, b) ((a) > (b)? (a): (b))

/**
 * O(n^2), Space: O(n)
 *
 * @param s
 * @return
 */
char *longestPalindrome(char *s) {
    if (s == NULL || strlen(s) == 0) {
        return "";
    }
    size_t n = strlen(s);
    int start = 0, maxLen = 0;
    int dp[n][n];
    for (int i = n - 1; i >= 0; i--) {
        for (int j = i; j < n; j++) {
            if (i == j) {
                dp[i][j] = 1;
            } else if (i + 1 == j) {
                dp[i][j] = s[i] == s[j];
            } else {
                dp[i][j] = s[i] == s[j] && dp[i + 1][j - 1];
            }
            if (dp[i][j] && j - i + 1 > maxLen) {
                maxLen = j - i + 1;
                start = i;
            }
        }
    }
    char *ret = malloc(maxLen * sizeof(char));
    int i;
    for (i = 0; start <= maxLen;) {
        ret[i++] = s[start++];
    }
    ret[i] = '\0';
    return ret;
}

/**
 *
 * @param s
 * @return
 */
char *longestPalindrome2(char *s) {
    int left, right, maxLen = 0, start = 0, idx = 0;
    while (s[idx]) {
        right = idx;
        left = idx - 1;
        while (s[right] == s[idx]) {
            right++;
        }
        idx = right;
        for (; left >= 0 && s[right] && s[right] == s[left]; left--, right++);
        if (right - left - 1 > maxLen) {
            start = left + 1;
            maxLen = right - left - 1;
        }
    }
    char *ret = (char *) malloc(maxLen + 1);
    for (int i = 0; i < maxLen; ++i) {
        ret[i] = s[start + i];
    }
    ret[maxLen] = '\0';
    return ret;
}

/**
 *
 * @param s
 * @param left
 * @param right
 * @return
 */
int expand(char *s, int left, int right) {
    for (; left >= 0 && s[right] && s[left] == s[right]; left--, right++);
    return right - left - 1;
}

/**
 *
 * @param s
 * @return
 */
char *longestPalindrome3(char *s) {
    if (s == NULL || strlen(s) == 0) {
        return "";
    }
    int start = 0, maxLen = 0, curLen;
    for (int i = 0; i < strlen(s); ++i) {
        curLen = max(expand(s, i, i), expand(s, i, i + 1));
        if (curLen > maxLen) {
            maxLen = curLen;
            start = i - (curLen - 1) / 2;
        }
    }
    char *ret = (char *) malloc(maxLen + 1);
    for (int i = 0; i < maxLen; ++i) {
        ret[i] = s[start + i];
    }
    ret[maxLen] = '\0';
    return ret;
}

int main() {
    printf("%s", longestPalindrome3("babad"));
    return 0;
}

File: go/src/problem/leetcode5.go
package leetcode

func longestPalindrome(s string) string {
    if len(s) == 0 {
        return s
    }
    expand := func(l, r int) int {
        for l >= 0 && r < len(s) && s[l] == s[r] {
            l--
            r++
        }
        return r - l - 1
    }
    start, maxLen := 0, 0
    for i := 0; i < len(s); i++ {
        l1, l2 := expand(i, i), expand(i, i+1)
        curLen := l1
        if l2 > l1 {
            curLen = l2
        }
        if curLen > maxLen {
            maxLen = curLen
            start = i - (curLen-1)/2
        }
    }
    return s[start : start+maxLen]
}

func longestPalindromeDP(s string) string {
    n, start, maxLen := len(s), 0, 0
    dp := make([][]bool, n)
    for i := 0; i < n; i++ {
        dp[i] = make([]bool, n)
    }
    for i := n - 1; i >= 0; i-- {
        for j := i; j < n; j++ {
            if i == j {
                dp[i][j] = true
            } else if i+1 == j {
                dp[i][j] = s[i] == s[j]
            } else {
                dp[i][j] = s[i] == s[j] && dp[i+1][j-1]
            }
            if dp[i][j] && j-i+1 > maxLen {
                maxLen = j - i + 1
                start = i
            }
        }
    }
    return s[start : start+maxLen]
}

func longestPalindromeDP2(s string) string {
    res, dp := "", make([][]bool, len(s))
    for i := 0; i < len(s); i++ {
        dp[i] = make([]bool, len(s))
    }
    for i := len(s) - 1; i >= 0; i-- {
        for j := i; j < len(s); j++ {
            dp[i][j] = (s[i] == s[j]) && ((j-i < 3) || dp[i+1][j-1])
            if dp[i][j] && (res == "" || j-i+1 > len(res)) {
                res = s[i : j+1]
            }
        }
    }
    return res
}
